AWS 'under the hood' document

justinsb opened the pull request:
Edit: I don't think WIP any longer, but comments still very welcome (and this is definitely v1).

This is a WIP PR to allow people to comment - comments are very welcome!

A lot of people have asked for a "what is kube-up" doing document, which describes the internals of what kube-up creates.   The intention is that this document will be sufficiently detailed that somebody could write a replacement for kube-up, or bring up kubernetes on AWS manually (with a lot of work, admittedly), or make sure that our CloudFormation templates are doing the same thing.

This is my first draft, very WIP document, mostly so I can provide a place for people to provide feedback.  Please feel free to add lots of questions and requests for more detail.

(This has also proved very helpful to me in identifying some of the remaining functionality gaps, which is why I just opened that series of AWS issues!)

justinsb committed:
AWS "under the hood" document

Document how we implement kubernetes on AWS, so that configuration tools
other than kube-up can have a reference for what they should do, and
generally to help developers get up to speed.

iterion commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,120 @@
+
+* Instances run Ubuntu 15.04.  It includes a sufficiently modern kernel to give
+  a good experience with Docker, it doesn't require a reboot.
+* By default we run btrfs as the filesystem on the nodes.  It is relatively
```
For now it looks as if we're using aufs by default:
https://github.com/GoogleCloudPlatform/kubernetes/blob/f48543aba54e7edd545389fd104695463f0a354b/cluster/aws/config-default.sh#L30

Perhaps you have a PR switching to btrfs I'm not seeing?

justinsb replied:
You're right thanks - will update

andrewhampton commented:
Hey @justinsb, this looks like a great start. We discussed this via email, but just to list them here for discussion, here are the 4 questions I would like this document to answer:

1. What AWS things does kube-up create?
2. What is the purpose of each of those things?
3. What are the details on the ec2 instances created? (e.g. what ami? what's on the ami? what's installed on them before salt takes over?)
4. What exactly does salt do?

So far, this document is doing a good job of answer #1 and #2. It touches briefly on #3 in the Top-level section, and it doesn't really address #4 at all. Here are a few follow-up questions that I hope help flesh it out.

1. Are the things salt does AWS specific? If not, they might not belong in this document, but it would be nice to see that documented somewhere. Also, it would at least be nice to know how salt gets bootstrapped and kicked off.
2. More details on how the ec2 instances are bootstrapped would be nice to see.
3. In the NodePort & LoadBalancer section, you say, "kube-proxy listens externally on that port on every minion". Either this isn't correct or I'm not understanding what's happening. I spun up a new examples/guestbook with the frontend-service using LoadBalancer. `kubectl describe service frontend` shows `NodePort:		<unnamed>	32309/TCP`, but I'm not able to hit that port to get to the app. I tried both from my local machine and via curl from one of the minions. I do see this in the iptables: `DNAT       tcp  --  anywhere             anywhere             /* default/frontend: */ tcp dpt:32309 to:172.20.0.238:52087`, but it doesn't seem to be working. I don't know a ton about iptables though, so I'm betting I'm just missing something.
4. Should mention the ssh key created and what user to use for ssh access

andrewhampton commented:
Also, it would be nice to see how to add configuration to $HOME/.dockercfg for accessing private repositories as described here:

https://github.com/GoogleCloudPlatform/kubernetes/blob/v1.0.1/docs/images.md#configuring-nodes-to-authenticate-to-a-private-repository

waltisfrozen commented:
Thanks @justinsb! Like @andrewhampton, I think this is a good start, but from the operations perspective, this doesn't really give me enough to manually configure my own Kubernetes cluster or troubleshoot when things go wrong.   

Looking at the documentation at kubernetes.io, the contributed example of cloud-init scripts using CoreOS gets much closer to what I'm looking for since I can look at the script and see exactly what's being installed, but it still falls short for a few reasons :

1) It installs an older version of k8s. Incrementing the version numbers in the script causes the build to break for some unknown reason. 

2) It doesn't get to the same place as the kube-up procedure. The CoreOS process ends with creating an ssh tunnel, but kube-up configures kubectl locally and creates a secure key for communicating with the master. How can I do this manually?

3) Like kube-up, it doesn't provide any guidance on how to modify or troubleshoot the components on my own. Can I install K8s in my existing VPC? How do I configure security groups? Can I still have SSL terminate on the ELB? Where are the logs?

If I'm someone who needs to support a Kubernetes infrastructure in AWS (or any other non-Google hosting environment), I need to know what's being installed on each instance, how the components interact, where the logs for each component are located, the monitoring endpoints (if any) for each component, what needs to be created beyond the individual instances, etc.

I'm *really* not a fan of the curl piped through bash approach to installing applications, so having this be the only "supported" procedure for installing/configuring a master, multiple-node minion cluster, and host of supporting infrastructure (ELB's, autoscaling, VPC, etc.) is troublesome. It's a black box, which is great for developers looking for a PaaS-like place to run code, but far from ideal for operations staff who are tasked with building and maintaining the underlying environments.

justinsb commented:
Great feedback - thanks @andrewhampton and @waltisfrozen 

I think I've added lots more details to address much of the feedback.

I've also found some other documents which I think are relevant (I don't want to document non-AWS specific things here):

Salt: https://github.com/kubernetes/kubernetes/blob/master/docs/admin/salt.md
Troubleshooting: https://github.com/kubernetes/kubernetes/blob/master/docs/admin/cluster-troubleshooting.md

In general, the whole docs/admin tree looks great for operational issues.  There is definitely room for improvement there (as there is here), so please keep filing issues / comment here :-)

@andrewhampton Your NodePort issue sounds like you do have the right understanding.  Perhaps open a separate issue for that if it isn't working as intended.  Note that on AWS we open e.g. a LoadBalancer in the firewall, but we don't open a NodePort, so it might be an AWS firewall issue.  Maybe double-check the `kubernetes-minion-<clusterid>` security group.

Also, I split off the private repository suggestion (which is a good one) into a new issue #12794 as I don't think it'll belong in this document.

@waltisfrozen Thanks for explaining your viewpoint!  I was really focused on the kube-up script, and I haven't really invested the time yet to get alternative options to parity.  But - particularly before this document - a bash script isn't the most opaque way to configure things!  Even after this document, people will still want to do things not through the script, and I think that's a good thing, though it will be hard until we make more progress on moving more logic out of the script and into k8s itself (I believe there's a long-term goal to have the scripts just provide bare instances and have one code path "take it from there").  I'm going to look at the cloudformation templates next, as I think it is just a matter of making sure that we have the right tags and they'll be much closer.  (Are these the same as the cloud-init setup you're using?  If not can you point me to the cloud-init docs?)

Going to remove the WIP tag so we can start the process of getting this merged.  I don't think this is perfect or entirely complete, but I hope it's a good start!

andrewhampton commented:
@justinsb regarding pulling from private repositories, I ended up using a secret instead. I followed the instructions here: https://github.com/kubernetes/kubernetes/blob/v1.0.1/docs/images.md#specifying-imagepullsecrets-on-a-pod

I actually like this solution better than adding the .dockercfg to each server since it's one less thing to worry about when adding a new minion to the cluster.

roberthbailey commented:
Assigning to @bgrant0607 to do a final pass on the docs.

waltisfrozen commented:
@justinsb Regarding the cloud-init scripts, I'm specifically referring to the "AWS CLI" section of the document "Getting started on Amazon EC2 with CoreOS" :

http://kubernetes.io/v1.0/docs/getting-started-guides/aws-coreos.html

The reasons I like this procedure more than the alternatives (kube-up, cloudformation) :

1) This process launches the master and worker nodes separately. It's great to know what a K8s environment looks like, how to configure workers, which pieces can be autoscaled, etc. 

2) Cloud-init has much more straighforward syntax than Bash or Cloudformation. I can also directly import the cloud-init scripts into the user data of an autoscaling config, which is nice.

3) Using the AWS CLI directly allows us to more easily make minor changes to the environment (eg. change instance size, lock down security group, etc.)

But, as I mentioned in my earlier comment, this process does fall short in a few areas : 

1) It doesn't touch on configuring kubectl. The ssh tunnel thing is a hack. 

2) It doesn't really delve into what each cloud-init script is doing. It mentions flannel and kube-register in the Highlights section, but leaves it there. 

3) It installs an older version of K8S.

4) It warns that it provisions K8s with "an insecure API server", but there's no link to a process for actually securing the API. 

My preference would be for this document to be the foundation for the "Getting started on AWS EC2" supported procedure and that it be updated to fill in the deltas between the vanilla installation and what's installed vis a vis the get-kube process.

andrewhampton commented:
FWIW, I agree with the suggestions made by @waltisfrozen.

bgrant0607 commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Peeking under the hood of kubernetes on AWS
```
Please capitalize Kubernetes

bgrant0607 replied:
everywhere in the doc

bgrant0607 replied:
Should be one `#` here

bgrant0607 commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+means that AWS will relaunch any minions that are terminated.
+
+We do not currently run the master in an AutoScalingGroup, but we should
+(#11934)
```
need period

bgrant0607 commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+### AutoScaling
+
+We run the minions in an AutoScalingGroup.  Currently auto-scaling (e.g. based
+on CPU) is not actually enabled (#11935).  Instead, the auto-scaling group
```
Issues are not auto-linked in docs. Please use the redirector:

`[#11935](http://issues.k8s.io/11935)`
and prs.k8s.io for PRs

bgrant0607 commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+### IAM
+
+kube-proxy sets up two IAM roles, one for the master called
+(kubernetes-master)[cluster/aws/templates/iam/kubernetes-master-policy.json]
```
Preview the doc. Link syntax is backward. Also please use working relative paths: ../../cluster/...

bgrant0607 commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+KubernetesCluster tag with a unique per-cluster value.
+
+
+# AWS Objects
```
Why is this a top-level section?

bgrant0607 commented:
Thanks for writing this up. Looks reasonable at a glance. Just some formatting issues. Reassigning to @RichieEscarez for final review.

RichieEscarez commented:
Starting my review now.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+## Peeking under the hood of kubernetes on AWS
+
+We encourage you to use kube-up (or CloudFormation) to create a cluster.  But
```
Edit (insert links):
`[kube-up](../getting-started-guides/aws.md)` 
`[CloudFormation](../getting-started-guides/aws-coreos.md)`

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+(please file an issue and let us know what's missing if there's something you
+need) but sometimes it is the only option.
+
+This document sets out to document how kubernetes on AWS maps to AWS objects.
```
Delete this line (moved above)

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+need) but sometimes it is the only option.
+
+This document sets out to document how kubernetes on AWS maps to AWS objects.
+Familiarity with AWS is assumed.
```
Delete this line (moved above)

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Peeking under the hood of kubernetes on AWS
+
```
Insert at top:

This document provides high-level insight into how Kubernetes works on AWS and maps to AWS objects. We assume that you are familiar with AWS.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+This document sets out to document how kubernetes on AWS maps to AWS objects.
+Familiarity with AWS is assumed.
+
+### Top-level
```
Replace with: 
### Architecture overview

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+### Top-level
+
+Kubernetes consists of a single master node, and a collection of minion nodes.
```
Edit (lines 48 - 51):
Kubernetes is a cluster of several machines that consists of a Kubernetes master and a set number of nodes (previously known as 'minions') for which the master which is responsible. See the `[Architecture](architecture.md)' topic for more details.

RichieEscarez replied:
Question: Are we able to avoid using the term "minion" in this doc? The correct term now is "node" but i see that there is still other references in the code like: `kubernetes-minion-<clusterid>`.

justinsb replied:
I'll try to avoid minion where it is english, but changing the code is out of scope for this PR :-)

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+* Instances run Ubuntu 15.04 (the official AMI).  It includes a sufficiently
+  modern kernel to give a good experience with Docker, it doesn't require a
+  reboot.  (The default SSH user is `ubuntu` for this and other ubuntu images)
```
Add a period at the end

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+By default on AWS:
+
+* Instances run Ubuntu 15.04 (the official AMI).  It includes a sufficiently
+  modern kernel to give a good experience with Docker, it doesn't require a
```
Revise to:
... modern kernel that pairs well with Docker and doesn't require a...

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+* By default we run aufs over ext4 as the filesystem / container storage on the
+  nodes (mostly because this is what GCE uses).
+
+These defaults can be changed by passing different environment variables to
```
Edit/Insert in what file:

These defaults can be changed by editing the `[config-default.sh](../../cluster/aws/config-default.sh)` file to specify what environment variables to pass to kube-up.

RichieEscarez replied:
Question: Is "config-default.sh" the right file that needs to be edited to pass in different env. variables?

justinsb replied:
Ah ... I don't think you're supposed to do that.  (I haven't, at least!)  Rather, I think you pass additional env vars to kube-up, and config-default will only set defaults if not otherwise set - at least for the variables that you're allowed to override!

justinsb replied:
I'm changing to "You can override these defaults by passing different environment variables to
kube-up."

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+### Storage
+
+AWS does support persistent volumes via EBS.  These can then be attached to
```
Edit/revise first sentence:
AWS supports persistent volumes by using `[Elastic Block Store (EBS)](../user-guide/volumes.md#awselasticblockstore)`.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+AWS does support persistent volumes via EBS.  These can then be attached to
+pods that should store persistent data (e.g. if you're running a database).
+
+Minions do not have persistent volumes otherwise.  In general, kubernetes
```
Edit (lines 69-71):

By default, nodes in AWS use `[instance storage](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/InstanceStorage.html)' unless you create pods with persistent volumes `[(EBS)](../user-guide/volumes.md#awselasticblockstore)`.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+containers do not have persistent storage unless you attach a persistent
+volume, and so minions on AWS use instance storage.  Instance storage is
+cheaper, often faster, and historically more reliable.  This does mean that you
+should pick an instance type that has sufficient instance storage, unless you
```
Edit for this whole sentence (starting at the end of line 72 "This does mean that you):

If you choose to use instance storage, you must select an instance type......

justinsb replied:
Although everyone uses instance storage: our containers are stored on instance storage.  If you are running e.g. mostly webservers you can probably make do with not a lot of disk because your is image is probably pretty small and you're not going to be creating a lot of data on disks.  But if either of those aren't true, you have to choose an instance type that includes more instance storage.

Right now the big gotcha here is that elasticsearch & influxdb create their data on instance storage.  They shouldn't IMHO, but they do...

RichieEscarez replied:
Thanks for clarifying, maybe we can rearrange that sentence to:

"Unless you can make do with whatever space is left on your root partition, you must choose an instance type that provides you with sufficient instance storage for your needs."

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+should pick an instance type that has sufficient instance storage, unless you
+can make do with whatever space is left on your root partition.
+
+The master _does_ have a persistent volume attached to it.  Containers are
```
Edit/revise to:

Note: Master uses a persistent volume (`[etcd](architecture.html#etcd)`) to track its state but similar to the nodes, container are mostly run against instance storage, except...

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+mostly run against instance storage, just like the minions, except that we
+repoint some important data onto the peristent volume.
+
+By default we use aufs over ext4.  `DOCKER_STORAGE=btrfs` is also a good choice
```
Edit (lines 80 - 82):

The default storage driver in the Docker images is aufs. Setting `DOCKER_STORAGE=btrfs` (in the  [config-default.sh](../../cluster/aws/config-default.sh) file) is also a good choic for a file system. btrfs is relatively reliable with Docker and has improved its reliability with modern kernels.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+### AutoScaling
+
+We run the minions in an AutoScalingGroup.  Currently auto-scaling (e.g. based
```
Edit (replace first line):

Nodes are run in an `[AutoScalingGroup](http://docs.aws.amazon.com/AutoScaling/latest/DeveloperGuide/AutoScalingGroup.html) on AWS.

justinsb replied:
Is there a way to make the master/minion distinction here?  Minions are run in an ASG, but the master currently is not.  I want to move the master to an ASG of size 1 so that it auto-recovers, but the fact that the master will not automatically recover is currently an important gotcha (that I need to fix)

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+### Networking
+
+Kubernetes uses an IP-per-pod model.  This means that a node, which runs many
+pods, must have many IPs.  The way we implement this on AWS is to use VPCs and
```
Edit (starting with "The way we implement..." through lines 98-99):

AWS uses VPCs and advanced routing support so each pod is assigned a /24 CIDR;....

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+Kubernetes uses an IP-per-pod model.  This means that a node, which runs many
+pods, must have many IPs.  The way we implement this on AWS is to use VPCs and
+the advanced routing support that it allows.  Each pod is assigned a /24 CIDR;
```
Delete (replace with edit above)

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+the advanced routing support that it allows.  Each pod is assigned a /24 CIDR;
+then this CIDR is configured to route to an instance in the VPC routing table.
+
+It is also possible to use overlay networking on AWS, but the default kube-up
```
Keep first part but edit after the comma,: 
…., but that is not the default configuration of the kube-up script.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+Kubernetes uses an IP-per-pod model.  This means that a node, which runs many
+pods, must have many IPs.  The way we implement this on AWS is to use VPCs and
+the advanced routing support that it allows.  Each pod is assigned a /24 CIDR;
+then this CIDR is configured to route to an instance in the VPC routing table.
```
Edit:

The assigned CIDR is then configured to....

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+### NodePort & LoadBalancing
+
+Kubernetes on AWS integrates with ELB.  When you create a service with
```
Edit - Replace "ELB" with:

`[Elastic Load Balancing (ELB)](http://docs.aws.amazon.com/AutoScaling/latest/DeveloperGuide/US_SetUpASLBApp.html)`

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+### NodePort & LoadBalancing
+
+Kubernetes on AWS integrates with ELB.  When you create a service with
+Type=LoadBalancer, kubernetes (the kube-controller-manager) will create an ELB,
```
Add code formatting:
`Type=LoadBalancer`

Also as mentioned in another comment, capital "K" in Kubernetes

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+minions to allow traffic from the ELB to the minions.  This traffic reaches
+kube-proxy where it is then forwarded to the pods.
+
+ELB requires that all minions listen on a single port, and it acts as a layer-7
```
Edit (lines 114 - 123):

ELB requires that all nodes listen on a single port and also act as a layer-7 forwarding proxy (i.e. the source IP is not preserved).  
In Kubernetes, `[LoadBalancer services](../user-guide/services.html#type-loadbalancer)` are exposed as `[NodePort services](../user-guide/services.html#type-nodeport)`. Then kube-proxy listens externally on the cluster-wide port that’s assigned to NodePort services and forwards traffic to the corresponding pods.  
So for ELB (as a load-balanced service), it’s configured to proxy traffic on the public port (e.g. port 80) to the NodePort that is assigned to the service (e.g. 31234). Any in-coming traffic to the NodePort (e.g. port 80) is recognized by kube-proxy and then sent to the correct pods (e.g pods with the 31234 service).

RichieEscarez replied:
Question: I edited this quite a bit, can you please ensure its technical accuracy?

justinsb replied:
Definitely an improvement, but in my original I had been ambiguous as to the subjects of some of my run-on sentences.  So I had to re-edit to get the right subjects in some places!

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+will likely have to open the port in the minion security group
+(`kubernetes-minion-<clusterid>`).
+
+### IAM
```
Replace:

### Identity and Access Management (IAM)

Also see my comment/question above about minions vs. nodes (there are several instances that can be replaced).

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+The minion policy is relatively minimal.  The master policy is probably overly
+permissive.  The security concious may want to lock-down the IAM policies
+further (#11936)
```
Same comment as above.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+further (#11936)
+
+We should make it easier to extend IAM permissions and also ensure that they
+are correctly configured (#???)
```
Just a reminder that this place-holder is here. #???

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+The minion does not need a lot of access to the AWS APIs.  It needs to download
+a distribution file, and then it is responsible for attaching and detaching EBS
+volumes to itself.
```
Edit:
...volumes from itself.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+### Tagging
+
+All AWS resources are tagged with a tag named "KuberentesCluster".  This tag is
```
Shorten to:

All AWS resources are tagged with "KubernetesCluster".

justinsb replied:
AWS tags are keys and values.  The name is "KubernetesCluster", with value is whatever per-cluster unique ID you have assigned.  (Things  are considered to belong to the same cluster if-and-only-if they have the the same KubernetesCluster tag value)

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+All AWS resources are tagged with a tag named "KuberentesCluster".  This tag is
+used to identify a particular 'instance' of Kubernetes, even if two clusters
+are deployed into the same VPC.  (The script doesn't do this by default, but it
```
Edit in the (..):
... (The kube-up script is not configured to do this by default but it is possible to create another cluster in the same VPC).

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+can be done.)
+
+Within the AWS cloud provider logic, we filter requests to the AWS APIs to
+match resources with our cluster tag.  So we only see our own AWS objects.
```
Revise last sentence:

... By filtering the requests, we ensure that we see only our AWS objects.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+Within the AWS cloud provider logic, we filter requests to the AWS APIs to
+match resources with our cluster tag.  So we only see our own AWS objects.
+
+If you choose not to use kube-up, you must tag everything with a
```
Edit/prefix with:

Important: 

And replace:

..., "you must tag everything with a KubernetesCluster tag with a unique per-cluster value."

To:

..., you must add another tag to everything that has a "KubernetesCluster" tag, so that it also has a unique per-cluster value.

justinsb replied:
I changed to:

If you choose not to use kube-up, you must pick a unique cluster-id value, and
ensure that all AWS resources have a tag with `Name=KubernetesCluster,Value=<clusterid>`.

This is the top thing that people that don't do kube-up get wrong, so I want to be really clear about it.  If you have a way to make it clearer, that is great!

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+The kube-up script does a number of things in AWS:
+
+* Creates an S3 bucket (`AWS_S3_BUCKET`) and copy the kubernetes distribution
```
Change "copy" to "copies"

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+* Creates an S3 bucket (`AWS_S3_BUCKET`) and copy the kubernetes distribution
+  and the salt scripts into it.  They are made world-readable and the HTTP URLs
+are passed to instances; this is how kubernetes code gets onto the machines.
+* Creates two IAM profiles based on templates in `cluster/aws/templates/iam`.
```
replace the period with a colon and then list the two templates
.... `cluster/aws/templates/iam`:

     * `kubernetes-master`: Used by the Kubernetes master. 
     * `kubernetes-minion`: Used by the nodes.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+by minion nodes.
+* Creates an AWS SSH key named `kubernetes-<fingerprint>`.  Fingerprint here is
+  the OpenSSH key fingerprint, so that multiple users can run the script with
+different keys and their keys will not collide (with near-certainty) It will
```
Missing a period

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+
+# AWS Objects
+
```
From here on, all list items should have a period.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+use an existing key if one is found at `AWS_SSH_KEY`, otherwise it will create
+one there.  (With the default ubuntu images, if you have to SSH in: the user is
+`ubuntu` and that user can `sudo`)
+* Creates a VPC for use with the cluster (with a CIDR of 172.20.0.0/16)., and
```
Remove both ., after the (..)

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+* Creates a route table for the VPC, with the internet gateway as the default
+  route
+* Creates a subnet (with a CIDR of 172.20.0.0/24) in the AZ `KUBE_AWS_ZONE`
+  (defaults to us-west-2a).  Currently kubernetes runs in a single AZ; there
```
Edit (replace with):
 * Creates a subnet (with a CIDR of 172.20.0.0/24) in the AZ KUBE_AWS_ZONE (defaults to us-west-2a). ***this first sentences is unchanged***  
***below has been edited and reformatted***  
Currently, each Kubernetes cluster runs in a single AZ on AWS. Although, there are two philosophies in discussion on how to achieve High Availability (HA):  
     * cluster-per-AZ: An independent cluster for each AZ, where each cluster is entirely separate. 
     * cross-AZ-clusters: A single cluster spans multiple AZs. The debate is open here, where cluster-per-AZ is discussed as more robust but cross-AZ-clusters are more convenient.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+* Associates the subnet to the route table
+* Creates security groups for the master node (`kubernetes-master-<clusterid>`)
+  and the minion nodes (`kubernetes-minion-<clusterid>`)
+* Configures security groups so that masters & minions can intercommunicate,
```
Edit (lines 200-202):

Configures security groups so that masters and nodes can communicate. This includes intercommunication between masters and nodes, opening SSH publicly for both masters and nodes, and opening port 443 on the master for the HTTPS API endpoints.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+  security group, IAM credentials etc.  An instance script is used to pass
+vital configuration information to Salt.  The hope is that over time we can
+reduce the amount of configuration information that must be passed in this way.
+* Once the instance is up, it attaches the EBS volume & sets up a manual
```
Replace '&' with 'and'

justinsb replied:
Done!

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+world on the master (for the HTTPS API endpoint)
+* Creates an EBS volume for the master node of size `MASTER_DISK_SIZE` and type
+  `MASTER_DISK_TYPE`
+* Launches a master node with a fixed IP address (172.20.0.9), with the
```
Edit (lines 205 - 208):

Launches a master node with a fixed IP address (172.20.0.9) that is also configure for the security group and all the necessary IAM credentials. An instance script is used to pass vital configuration information to Salt. Note: The hope is that over time we can reduce the amount of configuration information that must be passed in this way.

RichieEscarez replied:
Question: I removed "etc." since that was unclear. Is it correct the way that Ive edited it? Otherwise, can we add in the info that "etc." refers to?

justinsb replied:
There's a lot of extra configuration (the VPC, the subnet, the instance type etc), but I think you've got the non-obvious/important bits!

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+* Once the instance is up, it attaches the EBS volume & sets up a manual
+  routing rule for the internal network range (`MASTER_IP_RANGE`, defaults to
+10.246.0.0/24)
+* Creates an auto-scaling launch-configuration and group for the minions.  The
```
Edit (lines 212-220 + note the *italics* used for variable names):

For auto-scaling, on each nodes it creates a launch configuration and group. The name for both is <*KUBE_AWS_INSTANCE_PREFIX*>-minion-group. The default name is kubernetes-minion-group. The auto-scaling group has a min and max size that are both set to NUM_MINIONS. You can change the size of the auto-scaling group to add or remove the total number of nodes from within the AWS API or Console. Each nodes self-configures, meaning that they come up; run Salt with the stored configuration; connect to the master; are assigned an internal CIDR; and then the master configures the route-table with the assigned CIDR. The kube-up script performs a health-check on the nodes but it's a self-check that is not required.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+## Manual infrastructure creation
+
+While this work is not yet complete, advanced users may choose to create (some)
```
Edit (lines 232-233):
..., advanced users might choose to manually create AWS objects while still making use of the .....

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+
+While this work is not yet complete, advanced users may choose to create (some)
+AWS objects themselves, and still make use of the kube-up script (to configure
+Salt, for example).
```
Append (to introduce the list) + add periods to the end of each list item:

For AWS objects that you create, keep in mind the following:

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+  be reused
+* If your security groups are appropriately named, they will be reused.
+
+Currently there is no way to do the following with kube-up.  If these affect
```
Replace . period with a colon :

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+* If your security groups are appropriately named, they will be reused.
+
+Currently there is no way to do the following with kube-up.  If these affect
+you, please open an issue with a description of what you're trying to do (your
```
Delete lines 244 and 245 (see addition below).

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+you, please open an issue with a description of what you're trying to do (your
+use-case) and we'll see what we can do:
+
+* Use an existing AWS SSH key with an arbitrary name
```
A periods to each list item

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+use-case) and we'll see what we can do:
+
+* Use an existing AWS SSH key with an arbitrary name
+* Override the IAM credentials in a sensible way (but this is in-progress)
```
Edit in (...) - replace "but" with "note":
...(note that this is in-progress)

RichieEscarez replied:
Question: Is there an Issue # that we can include here?

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+* Override the IAM credentials in a sensible way (but this is in-progress)
+* Use different security group permissions
+* Configure your own auto-scaling groups
+
```
Insert here (move from above):

If any of the above items apply to your situation, open an issue to request an enhancement to the kube-up script. You should provide a complete description of the use-case, including all the details around what you want to accomplish.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,271 @@
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
```
Question: Will the "TOC script" be run on this fill to generate that TOC and list all the headers here?

justinsb replied:
I don't know a lot about how this works (new to writing docs, as you can probably tell).  If anyone can advise, that's great, otherwise I suspect we can fix it later....

RichieEscarez commented:
@justinsb, I made several edits for consistency and clarity as well as a few reformatting changes. Please let me know if any of my requested changes are not clear and also please help make sure they're all technically accurate.

justinsb committed:
Changes per reviews

justinsb committed:
Fix some typos from my read-through

justinsb commented:
Thanks for the reviews - as always @bgrant0607 but in particular @RichieEscarez - it is very obvious you're a pro-writer!  I don't know if this procedure is the most efficient way to do this (with your writing changes and me copying them in to the doc)  - I think if we had been sitting next to each other this could have been really fast, or I wouldn't have minded if you just took it and rewrote it!  Perhaps we can figure out something at the next design event (or even have a session for docs help).

I think it looks better (thank you!) and I think I have addressed all the comments.  I was a little confused on master vs minion but I think the minions are now "nodes" and the master is now "master", but the master is not considered a node.

I should probably squash the commits before we merge, but I won't do that until we're done with reviews!

justinsb committed:
Two small fixes (to keep doc-gen happy)

a-robinson commented:
@RichieEscarez - does this look good to you? It'd be nice to get merged.

RichieEscarez commented:
Thanks @a-robinson, Ill take a look (sorry I let this get lost in my inbox).

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+Kubernetes clusters are created on AWS. This can be particularly useful if
+problems arise or in circumstances where the provided scripts are lacking and
+you manually created or configured your cluster.
+
```
Lets make sure it gets included and manually insert the TOC here (note that it also includes some edits for clarify and capitalization consistency across the section titles (see my other comments))

```
**Table of contents:**
 * [Architecture overview](#architecture-overview)
 * [Storage](#storage)
 * [Auto Scaling group](#auto-scaling-group)
 * [Networking](#networking)
 * [NodePort and LoadBalancing services](#nodeport-and-loadbalancing-services)
 * [Identity and access management (IAM)](#identity-and-access-management-iam)
 * [Tagging](#tagging)
 * [AWS objects](#aws-objects)
 * [Manual infrastructure creation](#manual-infrastructure-creation)
 * [Instance boot](#instance-boot)
```

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+kernels.  It can easily span multiple volumes, which is particularly useful
+when we are using an instance type with multiple ephemeral instance disks.
+
+### AutoScaling
```
To match how AWS refers to it, lets revise this to:

Auto Scaling group

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+It is also possible to use overlay networking on AWS, but that is not the
+configuration of the kube-up script.
+
+### NodePort and LoadBalancing
```
For clarity, lets append 'services' (w/ lowercase 's') to this title:

NodePort and LoadBalancing services

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+value, and ensure that all AWS resources have a tag with
+`Name=KubernetesCluster,Value=<clusterid>`.
+
+### AWS Objects
```
Please change to lowercase 'o' for 'objects':

AWS objects

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+master which is responsible. See the [Architecture](architecture.md) topic for
+more details.
+
+Other documents describe the general architecture of Kubernetes (all nodes run
```
I dont think we need to mention this anymore since have the pointer to architecture.md above. Please delete lines 59-61.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+* Instances run Ubuntu 15.04 (the official AMI).  It includes a sufficiently
+  modern kernel that pairs well with Docker and doesn't require a
+  reboot.  (The default SSH user is `ubuntu` for this and other ubuntu images.)
+* By default we run aufs over ext4 as the filesystem / container storage on the
```
Lets cut out "default" here since we already say it in line 63. Please replace this bullet with:

` * Nodes use aufs instead of ext4 as the filesystem / container storage (mostly because this is what Google Compute Engine uses).`

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+[(EBS)](../user-guide/volumes.md#awselasticblockstore).  In general, Kubernetes
+containers do not have persistent storage unless you attach a persistent
+volume, and so nodes on AWS use instance storage.  Instance storage is cheaper,
+often faster, and historically more reliable.  This does mean that you should
```
I responded to an "outdated" comment thats not showing here so adding another comment here. Can we revise:
`This does mean that you should pick an instance type that has sufficient instance storage, unless you can make do with whatever space is left on your root partition.`

To:

"Unless you can make do with whatever space is left on your root partition, you must choose an instance type that provides you with sufficient instance storage for your needs."

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+do with whatever space is left on your root partition.
+
+Note: The master uses a persistent volume ([etcd](architecture.md#etcd)) to track
+its state but similar to the nodes, containers are mostly run against instance
```
Lets break this long sentence up. Please revise to:

"...to track its state. Similar to nodes, containers are also mostly...."

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+### AutoScaling
+
+Nodes (but not the master) are run in an
+[AutoScalingGroup](http://docs.aws.amazon.com/AutoScaling/latest/DeveloperGuide/AutoScalingGroup.html)
```
Also change this term to:

Auto Scaling group

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+Nodes (but not the master) are run in an
+[AutoScalingGroup](http://docs.aws.amazon.com/AutoScaling/latest/DeveloperGuide/AutoScalingGroup.html)
+on AWS.  Currently auto-scaling (e.g. based on CPU) is not actually enabled
+([#11935](http://issues.k8s.io/11935)).  Instead, the auto-scaling group means
```
Also revise "auto-scaling group" to:

Auto Scaling group

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+
+Kubernetes uses an IP-per-pod model.  This means that a node, which runs many
+pods, must have many IPs.  AWS uses virtual private clouds (VPCs) and advanced
+routing support so each pod is assigned a /24 CIDR.  Each pod is assigned a /24
```
Lets remove the first part of the sentence (that spans line 114 to 116, since its mentioned in the previous sentence). Please make the following changes:

Delete:
"Each pod is assigned a /24 CIDR;"

Then start the sentence as:
"The assigned CIDR is then......."

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+CIDR; the assigned CIDR is then configured to route to an instance in the VPC
+routing table.
+
+It is also possible to use overlay networking on AWS, but that is not the
```
Insert "default":

"....but that is not the default configuration of the..."

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+the nodes.  This traffic reaches kube-proxy where it is then forwarded to the
+pods.
+
+ELB has some restrictions: it requires that all nodes listen on a single port,
```
These restrictions are better highlighted if this was a bulleted list.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+[NodePort services](../user-guide/services.md#type-nodeport).  Then
+kube-proxy listens externally on the cluster-wide port that's assigned to
+NodePort services and forwards traffic to the corresponding pods.  So ELB is
+configured to proxy traffic on the public port (e.g. port 80) to the NodePort
```
Let me try to make this example clearer. Can you please make  a new paragraph and revise to:

So for example, if ELB is configured to proxy traffic on the public port 80 to the NodePort that is assigned to the service 31234. Then any in-coming traffic that is sent to that NodePort (e.g. port 31234) is recognized by kube-proxy and sent to the correct pods for that service.

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+ELB has some restrictions: it requires that all nodes listen on a single port,
+and it acts as a forwarding proxy (i.e. the source IP is not preserved).  To
+work with these restrictions, in Kubernetes, [LoadBalancer
+services](../user-guide/services.html#type-loadbalancer) are exposed as
```
Broken link: the ".html" should be ".md"

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+  the OpenSSH key fingerprint, so that multiple users can run the script with
+different keys and their keys will not collide (with near-certainty). It will
+use an existing key if one is found at `AWS_SSH_KEY`, otherwise it will create
+one there.  (With the default ubuntu images, if you have to SSH in: the user is
```
Capital 'U' for "...Ubuntu images..."

RichieEscarez commented on docs/design/aws_under_the_hood.md:
```diff
@@ -0,0 +1,302 @@
+  enables the `dns-support` and `dns-hostnames` options.
+* Creates an internet gateway for the VPC.
+* Creates a route table for the VPC, with the internet gateway as the default
+  route
```
For consistency, the bullets should all end with a period. (ill stop calling these out)

RichieEscarez commented:
@justinsb Thanks for you kind words! You did get the master/node changes right. Sorry i jumped right in and added more comments but nothing thats a real show stopper (i think i found 1 broken link). 

Let me know if I should just LGTM this to get it published/added to the repo and then I can take it, add the edits, and open another PR with the change in the file?

a-robinson commented:
I would just merge this for you now, but travis is failing due to a legitimate error. Run `hack/update-generated-docs.sh` to fix it.

RichieEscarez commented:
Once this gets merged, ill open another PR for the remaining edits.

wojtek-t commented:
It seems that failures are real failures:

/home/travis/gopath/src/github.com/kubernetes/kubernetes/docs/design/aws_under_the_hood.md
----
md-links:
contents were modified

@justinsb - can you please fix it?

dchen1107 commented:
https://github.com/kubernetes/kubernetes/pull/11938#issuecomment-147389532 is a real problem, not flaky test. I remove lgtm from pr so that other oncall person won't look into this.

justinsb committed:
More fixes based on commments

justinsb committed:
Remove broken link to CloudFormation setup

justinsb commented:
Fixed all the editorial comments (thanks again @RichieEscarez) and also fixed the build problem.

justinsb committed:
Rename LoadBalancing -> LoadBalancer

To match the Type value

justinsb commented:
@RichieEscarez does this look good now?

RichieEscarez commented:
Thanks! @justinsb LGTM (sorry for the delayed response)

a-robinson commented:
Thanks guys! Justin, mind squashing your commits down before merge?

justinsb commented:
Ooops... I was too slow on the squashing.  Sorry.
